examples/solidity/ERC20/erc20.qnt

```diff
@@ -1,246 +1,359 @@
-      nondet toAddr = oneOf(ADDR)
-      nondet value = oneOf(AMMOUNTS)
-      SubmitTransfer(sender, toAddr, value)
-    },
-    {
-      nondet sender = oneOf(ADDR)
-      nondet fromAddr = oneOf(ADDR)
-      nondet toAddr = oneOf(ADDR)
-      nondet value = oneOf(AMMOUNTS)
-      SubmitTransferFrom(sender, fromAddr, toAddr, value)
-    },
-    {
-      nondet sender = oneOf(ADDR)
-      nondet spender = oneOf(ADDR)
-      nondet value = oneOf(AMMOUNTS)
-      SubmitApprove(sender, spender, value)
-    },
-    {
-      nondet tx = oneOf(pendingTransactions)
-      any {
-        CommitTransfer(tx),
-        CommitTransferFrom(tx),
-        CommitApprove(tx),
-      }
-    },
-  }
-
-  /* False invariants to debug the spec */
-
-  /* Claim that no `TransferFrom` transaction is ever processed.
-    This is false. Use this false invariant to see an example of how
-    `TransferFrom` is processed. */
-
-  val NoTransferFrom =
-    val example = all {
-      lastTx.tag == "transferFrom",
-      not(lastTx.failed),
-      lastTx.value > 0,
-    }
-
-    not(example)
-
-
-  /* Claim that no `Approve` transaction is ever processed.
-    This is false. Use this false invariant to see an example of how
-    `Approve` is processed. */
-
-  val NoApprove =
-    val example = all {
-      lastTx.tag == "approve",
-      not(lastTx.failed),
-      lastTx.value > 0,
-    }
-
-    not(example)
-
-  /* EXPECTED PROPERTIES */
-
-  /* No transferFrom should be possible, while there is a pending approval
-    for a smaller amount. This invariant is violated, as explained in:
-    https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/edit# */
-
-  val NoTransferFromWhileApproveInFlight =
-    val badExample = all {
-      lastTx.tag == "transferFrom",
-      lastTx.value > 0,
-      not(lastTx.failed),
-      nondet tx = oneOf(pendingTransactions)
-      all {
-        tx.tag == "approve",
-        tx.sender == lastTx.fromAddr,
-        tx.spender == lastTx.sender,
-        tx.value < lastTx.value,
-        tx.value > 0,
-      }
-    }
-    not(badExample)
-
-  /* Make sure that the account balances never go negative. */
-  val NoNegativeBalances = ADDR.forall(a => balanceOf.get(a) >= 0)
+    }
+
+    /**
+     * ERC20: Moves amount tokens from the caller’s account to to.
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Quint: also return an error code and the new state.
+     * If the error != "", the new state should be applied.
+     */
+    pure def transfer(state: Erc20State, sender: Address,
+            toAddr: Address, amount: Uint): (bool, Erc20Result) = {
+        // `transfer` always returns true, but we should check Erc20Result.error
+        (true, _transfer(state, sender, toAddr, amount))
+    }
+
+    /**
+     * ERC20: Returns the remaining number of tokens that spender will be allowed to
+     * spend on behalf of owner through transferFrom. This is zero by default.
```

Comment:
This "zero by default" is from setting it in `newErc20`? Maybe you can explain it here.